mlkem/poly.c

```diff
@@ -367,6 +369,46 @@ void poly_getnoise_eta1(poly *r, const uint8_t seed[KYBER_SYMBYTES], uint8_t non
     poly_cbd_eta1(r, buf);
 }
 
+/*************************************************
+* Name:        poly_getnoise_eta1_4x
+*
+* Description: Batch sample four polynomials deterministically from a seed and nonces,
+*              with output polynomials close to centered binomial distribution
+*              with parameter KYBER_ETA1
+*
+* Arguments:   - poly *r{0,1,2,3}: pointer to output polynomial
+*              - const uint8_t *seed: pointer to input seed
+*                                     (of length KYBER_SYMBYTES bytes)
+*              - uint8_t nonce{0,1,2,3}: one-byte input nonce
+**************************************************/
+void poly_getnoise_eta1_4x(poly *r0,
+                           poly *r1,
+                           poly *r2,
+                           poly *r3,
+                           const uint8_t seed[KYBER_SYMBYTES],
+                           uint8_t nonce0,
+                           uint8_t nonce1,
+                           uint8_t nonce2,
+                           uint8_t nonce3)
+{
+    uint8_t buf[KECCAK_WAY][KYBER_ETA1 * KYBER_N / 4];
+    uint8_t extkey[KECCAK_WAY][KYBER_SYMBYTES + 1];
+    memcpy(extkey[0], seed, KYBER_SYMBYTES);
+    memcpy(extkey[1], seed, KYBER_SYMBYTES);
+    memcpy(extkey[2], seed, KYBER_SYMBYTES);
+    memcpy(extkey[3], seed, KYBER_SYMBYTES);
+    extkey[0][KYBER_SYMBYTES] = nonce0;
+    extkey[1][KYBER_SYMBYTES] = nonce1;
+    extkey[2][KYBER_SYMBYTES] = nonce2;
+    extkey[3][KYBER_SYMBYTES] = nonce3;
+    shake256x4(buf[0], buf[1], buf[2], buf[3], KYBER_ETA1 * KYBER_N / 4,
```

Comment:
There is a fair amount of unnecessary copying here which we should remove at some point, e.g. by specializing `shake256x4` to the relevant case where all input buffers are equal except for the last byte.

Reply:
Good idea, I add #88 so we don't forget in the future.